storybookjs/storybook
# Embed the preview context in the user's app

Issue: #3329

This is proof-of-concept of an idea to move the preview iframe inside the user's app.

The benefits of this are:
  1. The user does not need to configure webpack a second time for storybook.
  2. If the user's app doesn't run inside webpack, they can use storybook (e.g. Meteor apps, #3329)
  3. If they are running storybook alongside their app anyway, this avoids compiling changed components twice for every change.

The downsides I can see are:
 1. You need to change the code of your app to render the stories (And this might be a little tricky to get right so it doesn't affect your production bundle).
 2. You need to run both storybook and your app (which most do anyway I imagine).
 3. We lose the ability to control the way the stories are bundled, potentially hurting future plans to build small bundles for each storybook entrypoint (? cc @ndelangen)

I'm not necessarily suggesting we would *only* allow this way to do it, but just that we might offer it as an option for Meteor (& maybe other bundler) users, and potentially because it's an easier onboarding path for new storybook users.

## What I did

 - Changed the `@storybook/core` webpack config to not build the preview endpoint
 - Changed the `@storybook/core` provider to render http://localhost:3000/__storybook_preview__
 - Changed `@storybook/react` to (hackily) render inside an exported `StorybookPreview` component.

 - Changed the `cra-kitchen-sink` app to render the `StorybookPreview` if the url path is set to `__storybook_preview__`

 - Added a `meteor-react` example app that does the same.

## How to test

`yarn && yarn bootstrap --core` as usual.

### To test `storybook-cra`

Run `yarn storybook` and `yarn start`, browse to the storybook as usual. Notice the iframe points at the actual app.

### To test `meteor-react`

As meteor [doesn't work](https://github.com/meteor/meteor/issues/9000) with yarn workspaces, you'll need to `yarn link` both `@storybook/core` and `@storybook/react` into the app, then `yarn storybook` and run `meteor`, then browse to the storybook.

You can do:

```
meteor npm install --global yarn
meteor yarn
# after yarn linking in the relevant dirs
meteor yarn link @storybook/core
meteor yarn link @storybook/react

meteor yarn storybook
# and
meteor
```
